sandbox/apparmor/notify/listener/listener.go

```diff
@@ -416,21 +417,12 @@ func (l *Listener) handleRequestAaClassFile(buf []byte) error {
 	if err != nil {
 		return err
 	}
-	// Try to send the request over l.reqs. Do this in a separate goroutine
-	// so that it doesn't block if there is no reader.
-	l.tomb.Go(func() error {
-		select {
-		case l.reqs <- req:
-			// request received
-		case <-l.tomb.Dying():
-			// request not received, an automatic denial will be sent by waitAndRespond...
-			return l.tomb.Err()
-		}
-		return nil
-	})
-	// Since sending the req occurs in its own goroutine, we can run
-	// waitAndRespond... even if there is nothing reading requests, and if
-	// the listener starts dying, a denial is always sent to the kernel.
+	select {
```

Comment:
why stopping running this in a goroutine?

Reply:
A few reasons, though none are strictly necessary:
1. We don't need to read the next request from the kernel if there's not a listener ready to receive it --- nothing wrong with just letting the next request sit after the kernel writes it.
2. Sending requests synchronously guarantees their order.
3. We don't need to send an auto-deny on listener close (kernel does this for us) so no need to always call `waitAndRespondAaClassFile` even if the listener is closing before the request is received, and we can instead return early.
4. If we read in a request from the kernel, nobody reads it from `Reqs()` before it times out in the kernel (which sends auto-deny), and then something starts reading from `Reqs()`, that request will still be created, even though the operation has already been denied by the kernel --- this is unavoidable for the first request received from the kernel while there's no reader on `Reqs()`, but we can at least avoid it for any subsequent requests that the kernel might try to send while there is still no reader on `Reqs()`.
5. This reduces epoll syscalls to get data that we may not use, in the case that there's no reader on `Reqs()`: furthermore, the way `epoll.Wait()` is implemented internally, it sends repeated epoll requests with a (rather long) timeout so it can check if the epoll instance has been closed and stop retrying.
6. If a new listener is registered, the kernel will try to send requests to that listener instead, and in the future the kernel may re-issue past requests from listeners which were closed without responding to the new listener --- at the moment, though, this is not enabled, so we may as well read as few requests as is reasonable from the kernel so that they can be issues to future listeners rather than be responded to with an auto-deny from the kernel after the original listener disconnects.

I realize the last one in particular is a stretch. I think generally, there are a few slight-benefit reasons and only one minor drawbacks afaict.

The one drawback I could see is that if something finally starts reading from `Reqs()`, the listener has to do all the work to read subsequent requests from the kernel and create waiting goroutines starting at that point, when it could have instead been pre-doing the work before the reads to `Reqs()` begins. But that also would create a bunch of waiting goroutines running concurrently, which may never receive a reply and will only close when the listener closes.